[Ong Bing Jue] Duke Increments

bjhoohaha opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

Commit:
Merge branch 'gradle'

Integrate gradle support to Duke (Test)

Commit:
Level 1

-Implemented Greet, Echo, Exit

Commit:
Level-1

-Fixed bug in Duke.java
-Corrected title

Commit:
Level-2

-Add the ability to store whatever text entered by the user and display them back to the user when requested

Commit:
Level-3

-Add the ability to mark tasks as done.

Commit:
Level-4

-Add support for tracking three types of tasks:

ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

Commit:
Level-5

-Teach Duke to deal with errors such as incorrect inputs entered by the user.

Commit:
Level-6

-Add support for deleting tasks from the list.
-Rectified bug where deadlines were added instead of events

bjhoohaha committed:
-Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk whe Duke starts up.

bjhoohaha committed:
-Teach Duke to understand dates and times.

bjhoohaha committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  -Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk whe Duke starts up.

bjhoohaha committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  -Teach Duke to understand dates and times.

bjhoohaha committed:
Level-8

Teach Duke to understand dates and times.

bjhoohaha committed:
A-MoreOOP

-Refactor the code to extract out closely related code as classes.

bjhoohaha committed:
A-JUnit

Add JUnit tests to test the behavior of the code.

bjhoohaha committed:
A-Jar

Package the app as an executable JAR file so that it can be distributed easily.

bjhoohaha committed:
Refactor Duke for Readability

-Refactor Parser.java to individual Parsers e.g. CommandParser, DatePArser, FileLineParser,
-Handled exceptions that may occur

bjhoohaha committed:
Package Duke

Packaged java files into Duke into its constitutents.

Relevant package subdirectories are
-duke.command
-duke.exception
-duke.formatter
-duke.parser
-duke.storage
-duke.task
-duke.ui

bjhoohaha committed:
Add Javadoc for Duke

Add Javadoc comments for Duke
-Refactor code in Storage
-Refactor code in ExitCommand
-Add FileInWriteException

bjhoohaha committed:
Add JUnit and Fix bug for CommandParser

-Add JUnit tests to test the behavior of the code.
-Fix bug for CommandParser
-Refactor data/duke.txt

bjhoohaha committed:
Fix spelling mistake in Javadocs

-Package the app as an executable JAR file so that it can be distributed easily.
-Fix spelling mistake in javadocs of Duke.java

bjhoohaha committed:
Add Checkstyle

-Tweak the code to comply with a chosen coding standard.

bjhoohaha committed:
feat find command

-Features a find command to find all matching tasks that matches a given keyword
-Fix spelling mistakes in Javadocs
-Updated Javadocs to include FindCommand

niqiukun commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * This is the task to be added by the add command.
+     */
+    protected Task task;
```
Hi, I like the fact that you have included JavaDoc for protected class variables :) It will really help reviewers to understand your code.

niqiukun commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,53 @@
+ * The <code>AddCommand</code> is the parent class of {@link AddTodoCommand}, {@link AddDeadlineCommand},
+ * {@link AddEventCommand}.
+ */
+public abstract class AddCommand implements Command {
```
The `AddCommand` Class is declared abstract but there is no abstract methods to be implemented. Maybe declare `execute` as abstract so that its child classes has to implement it?

niqiukun commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    public boolean isExit() {
+        return false;
+    }
```
Since all child classes of `AddCommand` are not exit command and `isExit` is implemented in the parent class, probably there is no need for the child classes to override this method.

niqiukun commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,37 @@
+ * deadline task into the list of tasks in {@link duke.task.TaskList}. The user interface will display the new deadline
+ * task that is added.
+ */
+public class AddDeadlineCommand extends AddCommand {
```
An alternative way of implementing this is to have only one class of `AddCommand` and store the type of event as a class variable. But implementing them as child classes is also fine to me :)

febee99 replied:
Separating each task type into child classes that inherit from AddCommand works well and could be clearer and more comprehensive to the reader, although there might be lots of code repetition. As @niqiukun suggested, you could store the type of event as a class variable. Alternatively, you could store a private Task variable. `private Task t;`and then assign `t = new Deadline/Todo/Event` in your code body afterwards.

niqiukun commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * Returns <code>true</code> if the command is an exit command and <code>false</code> otherwise.
+     * @return <code>true</code> if the command is an exit command and <code>false</code> otherwise.
+     */
+    boolean isExit();
```
Consider an alternative implementation: you can declare `Command` as an abstract class, and implement `isExit` as follows: 
```Java
public boolean isExit() {
    return this instanceof ExitCommand;
}
```
As a result, you don't need to reimplement this method in all its child classes.

bjhoohaha replied:
Thank you. I think this implementation is sweet and has implemented it in my duke!

niqiukun commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,57 @@
+     * @param index the index of the task to be deleted in the list of tasks
+     * @throws InvalidParameterException if the index of the task specified is not a number
+     */
+    public DeleteCommand(String index) {
```
An alternative implementation is let the `Parser` handle type conversion and have the integer as a paramater here.

niqiukun commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * This is the parent class of all <code>RuntimeExcpetion</code> that may occur in the Duke program.
+ */
+public abstract class DukeException extends RuntimeException {
```
If you are expecting to catch and resolve the `DukeException` in your code, consider extending `Exception` instead of `RuntimeException`. Extending from `Exception` will require you to indicate in each method that you are throwing this exception, which is a good code practice. See another explanation [here](https://stackoverflow.com/questions/19857008/extending-exception-runtimeexception-in-java).

niqiukun commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * @param ui the user interface
+     * @param storage the storage for the tasks
+     */
+    void execute(TaskList tasks, Ui ui, Storage storage);
```
If your `DukeException` extends from `Exception`, you will have to indicate `throws ...` here. Please see my comments in the `DukeException` class.

niqiukun commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,143 @@
+        System.out.println("     ➕  " + task);
+        System.out.println("     Now you have " + size + " tasks in the list.");
+        System.out.println("     ------------------------------------------------------------");
+    }
```
It is good to see that you have different methods for message display for different purposes :)

niqiukun commented on src/main/java/duke/exception/InvalidCommandException.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * This is a runtime exception that occur when the user enters an invalid command.
+ */
+public class InvalidCommandException extends DukeException {
```
It is good to see that you have multiple child classes for `DukeException` that handle different types of exceptions :)

niqiukun commented on src/main/java/duke/exception/InvalidCommandException.java:
```diff
@@ -0,0 +1,37 @@
+     * @return the invalid command
+     */
+    public String getInvalidCommand() {
+        return command;
```
Note that if you construct the class with `InvalidCommandException()`, `command` may not be initialised when you call this method. My suggestion is to initialise `command` as `""` when the parameter is `null`.

niqiukun commented on data/duke.txt:
```diff
@@ -0,0 +1,3 @@
+todo,false,read books
+deadline,false,eat,22/08/2019 1800
+event,true,cook,25/08/2019 1200
```
You may consider adding this file into your `.gitignore` to prevent it from being tracked by git.

niqiukun commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,81 @@
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (InvalidCommandException ice) {
+                ui.showError("Invalid command: " + ice.getInvalidCommand());
```
Consider placing the type of exception in the `Ui` class? You can display different message for different types of exceptions by method overloading.

niqiukun commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,60 @@
+     * @throws IOException if an I/O error occurs
+     */
+    public void save(Stream<String> stream) throws IOException {
+        BufferedWriter writer = new BufferedWriter(new FileWriter(filePath));
```
You can use the `FileWriter` to write file directly without the need for `BufferedWriter`. See the following example:
```Java
FileWriter writer = new FileWriter(filePath);
writer.write("This is one line\n");
writer.flush();
writer.close();
```

niqiukun commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = DateParser.parse(by);
```
An alternative way of implementation is to keep the type conversion inside the `Parser` class. Hence, the constructor can take `Date` type as parameter directly. The benefit is that all type conversion related exceptions can be caught in the `Parser` class instead of being thrown by constructors of different classes.

niqiukun commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     * Encodes the deadline task to be stored in the storage.
+     * @return a string representation of the encoded deadline task
+     */
+    public String encode() {
```
Since you used a separate class for decoding the format from files (`FileLineParser`), consider using a seperate class for encoding them as well.

niqiukun commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+     * @param index the index of the task to be removed
+     * @return a string representation of the task removed to be printed on a user interface.
+     */
+    public String delete(int index) {
```
I would suggest to catch the `IndexOutOfBoundsException` here and throws a new `DukeException` to be caught by upper level class. Because the upper level class, i.e. `DeleteCommand`, should not be implementation dependent and hence should not know that it has to catch `IndexOutOfBoundsException`.

niqiukun commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,57 @@
+        try {
+            String task = tasks.delete(index);
+            ui.showDeletedMessage(task, tasks.size());
+        } catch (IndexOutOfBoundsException aioube) {
```
I would suggest to catch `IndexOutOfBoundsException` in `TaskList` instead of here because the actual implementation (and hence the need to catch `IndexOutOfBoundsException`) is not revealed to this class. Please see my comments in `TaskList` class.

niqiukun commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+     * @param index the index of the task to be done.
+     * @return a string representation of the done task to be be printed on a user interface.
+     */
+    public String done(int index) {
```
Same comment as `delete(int index)` method.

niqiukun commented on src/test/java/duke/parser/DateParserTest.java:
```diff
@@ -0,0 +1,42 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+class DateParserTest {
```
An alternative way to test the parser is to compare the parsed result with an actual instance of `Date` object. See the example below:
```Java
Calendar calendar = Calendar.getInstance();
calendar.set(Calendar.DAY_OF_MONTH, 28);
calendar.set(Calendar.MONTH, Calendar.AUGUST);
calendar.set(Calendar.YEAR, 2019);
calendar.set(Calendar.HOUR_OF_DAY, 23);
calendar.set(Calendar.MINUTE, 59);
calendar.set(Calendar.SECOND, 0);
calendar.set(Calendar.MILLISECOND, 0);
Date date = calendar.getTime();
assertEquals(date, DateUtil.parseDate("28/8/2019 2359"));
```

niqiukun reviewed (commented):
Hi Bing Jue, your code is quite well written :) Keep up the good work! I have left several comments in the code with alternative implementations so that you may consider improving it further.

febee99 commented on docs/duke/ui/Ui.html:
```diff
@@ -0,0 +1,525 @@
+<!DOCTYPE HTML>
```
You can consider adding these files to .gitignore :)

febee99 commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,37 @@
+
+    /**
+     * Constructs a new add deadline command that will be executed in the <code>run</code> method of
+     * {@link duke.main.Duke} with the specified line as a parameter.
```
Perhaps try to leave an empty line after the description for every JavaDoc comment in adherance to the style guide.

febee99 commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,37 @@
+            String[] arr = super.line.split(" /at ");
+            super.task = new Event(arr[0], arr[1]);
+        } catch (ArrayIndexOutOfBoundsException aioobe) {
+            throw new InvalidParameterException(line.isBlank() ? null : line);
```
I like how you have individual classes that extend from DukeException! 👍 It makes things much clearer as it specifies exactly what type of exception is being thrown.

febee99 commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,58 @@
+
+    /**
+     * The index of the task to be mark as done in the list of tasks.
+     */
```
I like how you wrote JavaDoc comments for even private variables! 💯

febee99 commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,42 @@
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        try {
+            storage.save(tasks.save());
```
Good idea to only save the tasks after all commands have been ran, rather than saving and updating time after time.

febee99 commented on src/main/java/duke/parser/CommandParser.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    public static Command parse(String fullCommand) {
+        String[] arr = fullCommand.split(" ");
+        String parameter = Arrays.stream(arr).skip(1).collect(Collectors.joining(" ")).trim();
```
Sleek method to parse the command. 💯

febee99 commented on src/main/java/duke/parser/TaskParser.java:
```diff
@@ -0,0 +1,20 @@
+ */
+public class TaskParser {
+
+    /**
```
As this class only contains one method which is parse(), it could perhaps be done away with. You could also consider shifting the parse method into the Task class, where it can be used like `Task.parse()` without having to specify the task in the parameters. :)

febee99 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+     */
+    public TaskList(Stream<String> lines) {
+        this.taskList = new ArrayList<>();
+        lines.forEach(line -> {
```
Interesting use of Streams here

febee99 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,143 @@
+     * Prints the current tasks in the list.
+     * @param list the list of tasks to be printed
+     */
+    public void showTable(String list) {
```
Effort for your formatting 💯

febee99 commented on src/test/java/duke/parser/DateParserTest.java:
```diff
@@ -0,0 +1,42 @@
+        Assertions.assertThrows(InvalidDateTimeException.class, () -> DateParser.parse(""));
+        Assertions.assertThrows(InvalidDateTimeException.class, () -> DateParser.parse("11/02/2931"));
+        Assertions.assertThrows(InvalidDateTimeException.class, () -> DateParser.parse("11/02/2931 test"));
+        Assertions.assertThrows(NullPointerException.class, () -> DateParser.parse(null));
```
Perhaps you could standardise the exceptions thrown such that if the date is null, InvalidDateTimeException will also be thrown instead of NullPointerException

febee99 reviewed (commented):
Hi there, I really enjoyed reading and evaluating your work. Keep it up! I have added several comments in the files committed and I hope that they can come in useful in one way or another.

bjhoohaha committed:
Added JavaFx

Added JavaFx up to tutorial 3

bjhoohaha committed:
Added JavaFX-FXML

Integrated JavaFx-FXML into application

bjhoohaha committed:
Designed Duke Graphical User Interface

Designed a new graphical user interface to integrate with duke using JavaFx.

bjhoohaha committed:
Added Graphic User Interface

Integrated Duke with Graphic User Interface

bjhoohaha committed:
Refactor duke.view & duke.commands

Cleaned up code in duke.view, JavaFXM and duke.commands

bjhoohaha committed:
Added Welcome Message for GUI

bjhoohaha committed:
Refactor duke & Fix storage issues

-Cleaned up code
-Fixed bug where tasks could not be saved once original file in directory is deleted
-Used streams
-Used lambda expressions

bjhoohaha committed:
Generate new jar file

bjhoohaha committed:
Refactor Code & Fix Storage Issues

-Cleaned up code
-Fix storage issues where data could not be saved in original file in directory is deleted or lost.
-Generate new Jar File

bjhoohaha committed:
Merge pull request #5 from bjhoohaha/refactor-code

Refactor Code & Fix Storage Issues

bjhoohaha committed:
Cleaned java import statememts

-Removed unused import statements

bjhoohaha committed:
Delete duke.txt

bjhoohaha committed:
Added Reminders, Sorting, Schedule

-User can set reminders
-Alerts will popup to remind user
-Use of insertion sort to sort tasks in schedule
-Allow display of tasks by date

bjhoohaha committed:
Fixed compiler warning where type of List is not checked

bjhoohaha committed:
Fixed compiler warnings where type of List is not checked

bjhoohaha committed:
Refactor and validate parameters passed into 'schedule' command

bjhoohaha committed:
Added java assertions and fixed bug where task at wrong index is mark as done

bjhoohaha committed:
Merge pull request #6 from bjhoohaha/assertions

Added java assertions and fixed bug where task at wrong index is mark…

bjhoohaha committed:
Optimized Insertion Sort

bjhoohaha committed:
Added javadocs for some duke packages and added Ui.png

Added javadocs for duke.view
Added javadocs for duke.ui
Added javadocs for duke.task
Added javadocs for duke.storage
Added javadocs for duke.parser
Added javadocs for duke.main
Added javadocs for duke.formatter
Added Ui.png

bjhoohaha committed:
Added Javadocs and Checkstyle

bjhoohaha committed:
Credit reused work

bjhoohaha committed:
Update Storage.java

bjhoohaha committed:
Fix font issues in windows and linux distribution and give credit to author

bjhoohaha committed:
Credit to Fonts Designer

bjhoohaha committed:
Set theme jekyll-theme-architect

bjhoohaha committed:
Add pictures for readme

bjhoohaha committed:
Merge branch 'master' of https://github.com/bjhoohaha/duke

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Rename index.html

bjhoohaha committed:
Removed html, js files for javadocs

bjhoohaha committed:
Update docs/README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Set theme jekyll-theme-architect

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Merge branch 'master' of https://github.com/bjhoohaha/duke

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Fixed issue where multipler timers are set in the future and added demo images

bjhoohaha committed:
Renamed demo pictures

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update docs/README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Update README.md

bjhoohaha committed:
Release Duke

bjhoohaha committed:
Update binary file for release
